snakeviz/ipymagic.py

```diff
@@ -1,38 +1,120 @@
+            """
+            Profile code and display the profile in Snakeviz.
+            Works as a line or cell magic.
+
+            Usage, in line mode:
+            %snakeviz [options] statement
+
+            Usage, in cell mode:
+            %%snakeviz [options] [statement]
+            code...
+            code...
 
-    # generate the stats file using IPython's prun magic
-    ip = get_ipython()
+            Options:
 
-    if cell:
-        ip.run_cell_magic('prun', line, cell)
-    else:
-        ip.run_line_magic('prun', line)
+            -e/--embed
+            If running the snakeviz magic in the Jupyter Notebook,
+            use this flag to embed the snakeviz visualization within
+            the notebook instead of trying to open a new tab.
 
-    # start up a Snakeviz server
-    sv = subprocess.Popen(['snakeviz', filename])
+            """
+            # get location for saved profile
+            filename = tempfile.NamedTemporaryFile().name
 
-    # give time for the Snakeviz page to load then shut down the server
-    time.sleep(3)
-    sv.terminate()
+            # parse options
+            opts, line = self.parse_options(line, 'e', 'embed', posix=False)
+
+            # call signature for prun
+            line = '-q -D ' + filename + ' ' + line
+
+            # generate the stats file using IPython's prun magic
+            ip = get_ipython()
+
+            if cell:
+                ip.run_cell_magic('prun', line, cell)
+            else:
+                ip.run_line_magic('prun', line)
+
+            # start up a Snakeviz server
+            if _check_ipynb() and ('e' in opts or 'embed' in opts):
+                sv = open_snakeviz_and_display_in_notebook(filename)
+            else:
+                sv = subprocess.Popen(['snakeviz', filename])
+            # give time for the Snakeviz page to load then shut down the server
+            time.sleep(3)
+            sv.terminate()
 
 
 def load_ipython_extension(ipython):
-    ipython.register_magic_function(snakeviz_magic, magic_kind='line_cell',
-                                    magic_name='snakeviz')
+    """Called when user runs %load_ext snakeviz"""
+    ipython.register_magics(SnakevizMagic)
+
+
+def _check_ipynb():
+    """
+    Returns True if IPython is running as the backend for a
+    Jupyter Notebook.
+
+    """
+    cfg = get_ipython().config
+    return "connection_file" in cfg["IPKernelApp"]
+
+
+def open_snakeviz_and_display_in_notebook(filename):
+
+    def _find_free_port():
+        import socket
+        from contextlib import closing
+        with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
+            s.bind(('', 0))
+            return s.getsockname()[1]
+
+    port = str(_find_free_port())
+
+    def _start_and_wait_when_ready():
+        import os
+        environ = os.environ.copy()
+        environ["PYTHONUNBUFFERED"] = "TRUE"
+        sv = subprocess.Popen(
+            ['snakeviz', "-s", "-p", port, filename],
```

Comment:
We need 0.0.0.0
In my experience, me and my colleagues always run jupyter notebook on the server and open it in our own laptop. So we cannot force the iframe to look at localhost. And also its not always the case the ip we found in ipython configuration is the same ip we use to connect to the server, for example when we run the jupyter inside docker, the ip in jupyter is the container's ip but the ip we need to connect to is the host's ip. And that's why I put a JS script in iframe to use the same ip with the notebook.
For security concern, I think its fine because:
1. the port is random
2. the process will only alive for 3 seconds and after that it will be killed.

Reply:
@yjmade, Noted! I added back host `0.0.0.0` and added a note so people are aware of it.